File: sdks/java/src/test/java/dev/atap/sdk/crypto/DPoPProofTest.java
package dev.atap.sdk.crypto;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;

import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class DPoPProofTest {

    @Test
    void testCreateWithoutAccessToken() {
        Ed25519KeyPair kp = Ed25519KeyPair.generate();
        String proof = DPoPProof.create(kp, "POST", "https://localhost:8080/v1/oauth/token");

        assertNotNull(proof);
        String[] parts = proof.split("\\.");
        assertEquals(3, parts.length);
    }

    @Test
    void testCreateWithAccessToken() {
        Ed25519KeyPair kp = Ed25519KeyPair.generate();
        String proof = DPoPProof.create(kp, "GET", "https://localhost:8080/v1/entities/abc",
                "test_access_token");

        assertNotNull(proof);
        String[] parts = proof.split("\\.");
        assertEquals(3, parts.length);
    }

    @Test
    @SuppressWarnings("unchecked")
    void testHeaderContainsCorrectFields() throws Exception {
        Ed25519KeyPair kp = Ed25519KeyPair.generate();
        String proof = DPoPProof.create(kp, "POST", "https://example.com/token");

        String[] parts = proof.split("\\.");
        byte[] headerBytes = Base64Url.decode(parts[0]);
        ObjectMapper mapper = new ObjectMapper();
        Map<String, Object> header = mapper.readValue(headerBytes, Map.class);

        assertEquals("dpop+jwt", header.get("typ"));
        assertEquals("EdDSA", header.get("alg"));
        assertNotNull(header.get("jwk"));

        Map<String, Object> jwk = (Map<String, Object>) header.get("jwk");
        assertEquals("OKP", jwk.get("kty"));
        assertEquals("Ed25519", jwk.get("crv"));
        assertNotNull(jwk.get("x"));
    }

    @Test
    @SuppressWarnings("unchecked")
    void testPayloadContainsCorrectClaims() throws Exception {
        Ed25519KeyPair kp = Ed25519KeyPair.generate();
        String proof = DPoPProof.create(kp, "GET", "https://localhost:8080/v1/test");

        String[] parts = proof.split("\\.");
        byte[] payloadBytes = Base64Url.decode(parts[1]);
        ObjectMapper mapper = new ObjectMapper();
        Map<String, Object> payload = mapper.readValue(payloadBytes, Map.class);

        assertEquals("GET", payload.get("htm"));
        assertEquals("https://localhost:8080/v1/test", payload.get("htu"));
        assertNotNull(payload.get("iat"));
        assertNotNull(payload.get("jti"));
        assertNull(payload.get("ath")); // no access token
    }

    @Test
    @SuppressWarnings("unchecked")
    void testPayloadContainsAthWhenAccessTokenProvided() throws Exception {
        Ed25519KeyPair kp = Ed25519KeyPair.generate();
        String proof = DPoPProof.create(kp, "POST", "https://localhost:8080/v1/test",
                "my_access_token");

        String[] parts = proof.split("\\.");
        byte[] payloadBytes = Base64Url.decode(parts[1]);
        ObjectMapper mapper = new ObjectMapper();
        Map<String, Object> payload = mapper.readValue(payloadBytes, Map.class);

        assertNotNull(payload.get("ath"));
    }

    @Test
    void testSignatureIsVerifiable() {
        Ed25519KeyPair kp = Ed25519KeyPair.generate();
        String proof = DPoPProof.create(kp, "POST", "https://example.com/token");

        String[] parts = proof.split("\\.");
        byte[] signingInput = (parts[0] + "." + parts[1]).getBytes();
        byte[] signature = Base64Url.decode(parts[2]);

        assertTrue(kp.verify(signingInput, signature));
    }

    @Test
    void testEachProofHasUniqueJti() {
        Ed25519KeyPair kp = Ed25519KeyPair.generate();
        String proof1 = DPoPProof.create(kp, "POST", "https://example.com/token");
        String proof2 = DPoPProof.create(kp, "POST", "https://example.com/token");

        assertNotEquals(proof1, proof2);
    }
}

File: sdks/php/tests/Crypto/DPoPProofTest.php
<?php

declare(strict_types=1);

namespace Atap\Sdk\Tests\Crypto;

use Atap\Sdk\Crypto\Base64Url;
use Atap\Sdk\Crypto\DPoPProof;
use Atap\Sdk\Crypto\Ed25519;
use Atap\Sdk\Tests\TestCase;

class DPoPProofTest extends TestCase
{
    public function testCreateProducesThreeParts(): void
    {
        $keys = Ed25519::generateKeypair();
        $proof = DPoPProof::create($keys['secretKey'], 'POST', 'https://example.com/token');

        $parts = explode('.', $proof);
        $this->assertCount(3, $parts);
    }

    public function testHeaderHasCorrectFields(): void
    {
        $keys = Ed25519::generateKeypair();
        $proof = DPoPProof::create($keys['secretKey'], 'POST', 'https://example.com/token');

        $parts = explode('.', $proof);
        $header = json_decode(Base64Url::decode($parts[0]), true);

        $this->assertSame('dpop+jwt', $header['typ']);
        $this->assertSame('EdDSA', $header['alg']);
        $this->assertArrayHasKey('jwk', $header);
        $this->assertSame('OKP', $header['jwk']['kty']);
        $this->assertSame('Ed25519', $header['jwk']['crv']);
        $this->assertArrayHasKey('x', $header['jwk']);
    }

    public function testPayloadHasRequiredClaims(): void
    {
        $keys = Ed25519::generateKeypair();
        $proof = DPoPProof::create($keys['secretKey'], 'GET', 'https://example.com/resource');

        $parts = explode('.', $proof);
        $payload = json_decode(Base64Url::decode($parts[1]), true);

        $this->assertArrayHasKey('jti', $payload);
        $this->assertSame('GET', $payload['htm']);
        $this->assertSame('https://example.com/resource', $payload['htu']);
        $this->assertArrayHasKey('iat', $payload);
        $this->assertIsInt($payload['iat']);
    }

    public function testPayloadIncludesAthWhenAccessTokenProvided(): void
    {
        $keys = Ed25519::generateKeypair();
        $proof = DPoPProof::create(
            $keys['secretKey'],
            'POST',
            'https://example.com/resource',
            'my_access_token',
        );

        $parts = explode('.', $proof);
        $payload = json_decode(Base64Url::decode($parts[1]), true);

        $this->assertArrayHasKey('ath', $payload);

        // Verify ath is SHA-256 of access token
        $expectedAth = Base64Url::encode(hash('sha256', 'my_access_token', true));
        $this->assertSame($expectedAth, $payload['ath']);
    }

    public function testPayloadHasNoAthWithoutAccessToken(): void
    {
        $keys = Ed25519::generateKeypair();
        $proof = DPoPProof::create($keys['secretKey'], 'POST', 'https://example.com/token');

        $parts = explode('.', $proof);
        $payload = json_decode(Base64Url::decode($parts[1]), true);

        $this->assertArrayNotHasKey('ath', $payload);
    }

    public function testSignatureIsValid(): void
    {
        $keys = Ed25519::generateKeypair();
        $proof = DPoPProof::create($keys['secretKey'], 'POST', 'https://example.com/token');

        $parts = explode('.', $proof);
        $signingInput = $parts[0] . '.' . $parts[1];
        $signature = Base64Url::decode($parts[2]);

        $this->assertTrue(Ed25519::verify($signature, $signingInput, $keys['publicKey']));
    }

    public function testJtiIsUnique(): void
    {
        $keys = Ed25519::generateKeypair();
        $proof1 = DPoPProof::create($keys['secretKey'], 'POST', 'https://example.com/token');
        $proof2 = DPoPProof::create($keys['secretKey'], 'POST', 'https://example.com/token');

        $parts1 = explode('.', $proof1);
        $parts2 = explode('.', $proof2);
        $payload1 = json_decode(Base64Url::decode($parts1[1]), true);
        $payload2 = json_decode(Base64Url::decode($parts2[1]), true);

        $this->assertNotSame($payload1['jti'], $payload2['jti']);
    }

    public function testPublicKeyInHeaderMatchesSecretKey(): void
    {
        $keys = Ed25519::generateKeypair();
        $proof = DPoPProof::create($keys['secretKey'], 'POST', 'https://example.com/token');

        $parts = explode('.', $proof);
        $header = json_decode(Base64Url::decode($parts[0]), true);
        $xBytes = Base64Url::decode($header['jwk']['x']);

        $this->assertSame($keys['publicKey'], $xBytes);
    }
}
